src/Compilers/Test/Core/Assert/AssertEx.cs

```diff
@@ -594,37 +616,117 @@ public static void AssertResultsEqual(string result1, string result2)
+        {
+            expectedSubString = NormalizeWhitespace(expectedSubString, out var pooled1);
+            actualString = NormalizeWhitespace(actualString, out var pooled2);
+
+            if (!actualString.StartsWith(expectedSubString, StringComparison.Ordinal))
+            {
+                // Call through to actual API to get same error message experience
+                Assert.StartsWith(expectedSubString.ToString(), actualString.ToString(), StringComparison.Ordinal);
+            }
+
+            if (pooled1 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled1);
+            }
+            if (pooled2 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled2);
+            }
+        }
+
+        internal static ReadOnlySpan<char> NormalizeWhitespace(ReadOnlySpan<char> input, out char[] pooledAlloc)
+        {
+            return NormalizeWhitespaceROM(input, out pooledAlloc).Span;
         }
 
-        internal static string NormalizeWhitespace(string input)
+        internal static ReadOnlyMemory<char> NormalizeWhitespaceROM(ReadOnlySpan<char> input, out char[] pooledAlloc)
         {
-            var output = new StringBuilder();
-            var inputLines = input.Split('\n', '\r');
-            foreach (var line in inputLines)
+            // Setup allocation (start with none)
+            pooledAlloc = null;
+            int allocSize = 0;
+
+            // Loop over each line section until we run out of lines
+            int prev = 0;
+            int next = input.IndexOfAny("\r\n".AsSpan());
+            if (next < 0) next = input.Length;
+            while (true)
             {
-                var trimmedLine = line.Trim();
+                // Get the line, trimmed
+                var trimmedLine = input.Slice(prev, next - prev).Trim();
                 if (trimmedLine.Length > 0)
                 {
+                    // We want to append 2 spaces first if we don't start with { or }
+                    int spacesToAppend = 0;
                     if (!(trimmedLine[0] == '{' || trimmedLine[0] == '}'))
                     {
-                        output.Append("  ");
+                        spacesToAppend = 2;
                     }
 
-                    output.AppendLine(trimmedLine);
+                    // Ensure buffer size is big enough
+                    int newChars = spacesToAppend + trimmedLine.Length + Environment.NewLine.Length;
+                    pooledAlloc ??= ArrayPool<char>.Shared.Rent(newChars);
+                    if (pooledAlloc.Length < allocSize + newChars)
+                    {
+                        char[] newArr = ArrayPool<char>.Shared.Rent(allocSize + newChars);
+                        pooledAlloc.AsSpan().Slice(0, allocSize).CopyTo(newArr.AsSpan().Slice(0, allocSize));
+                        ArrayPool<char>.Shared.Return(pooledAlloc);
+                        pooledAlloc = newArr;
```

Comment:
This pattern of basically increasing the size of the pooled array is repeated a few times in the PR. Think we should factor this out into a helper method.

Reply:
Do you mean another inline method, or in some helper type? If you mean in a helper type, which one, or where would I create it?